core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -129,6 +133,36 @@ public TableMetadata refresh() {
 
   @Override
   public void commit(TableMetadata base, TableMetadata metadata) {
+    // findOldVersionHint-------------------->isFirstRun------->JobFail
```

Comment:
This seems very complicated and i'm not sure it is correct

Reply:
The original intent of this is to 
1. avoid as much as possible the situation where the versionHint does not correspond to the Metadata 
2. optimise concurrent commits in streaming and LLAP scenarios. If the versionHint does not exist, it is assumed that another client is committing, and the current client aborts the commit.